src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -1818,6 +1825,145 @@ impl Coordinator {
+            plan.finishing,
+            select_id,
+            index_id,
+            optimizer_config.clone(),
+        );
+        let local_plan = optimizer.optimize(plan.source)?;
+        let resolved = local_plan.resolve(ctx.session(), oracle);
+        let global_plan = optimizer.optimize(resolved)?;
+
+        // We have the sources (including MFPs) and the corresponding timestamps; collect stats.
+        for (gid, (desc, _)) in global_plan.df_desc().source_imports.iter() {
+            let catalog_entry = self.catalog.get_entry(gid);
+            let full_name = self
+                .catalog
+                .for_session(&ctx.session)
+                .resolve_full_name(&catalog_entry.name);
+            let name = format!("{}", full_name);
+            let relation_desc = catalog_entry
+                .item
+                .desc_opt()
+                .expect("source should have a proper desc");
+            let stats: SnapshotStats = self
+                .controller
+                .storage
+                .snapshot_stats(*gid, as_of.clone())
+                .await?;
+
+            let mut total_bytes = 0;
+            let mut total_parts = 0;
+            let mut selected_bytes = 0;
+            let mut selected_parts = 0;
+            for (bytes, stats) in &stats.parts {
+                let bytes = u64::cast_from(*bytes);
+                total_bytes += bytes;
+                total_parts += 1u64;
+                let selected = match stats {
+                    None => true,
+                    Some(stats) => {
+                        let stats = stats.decode();
+                        let stats = RelationPartStats::new(name.as_str(), &relation_desc, &stats);
+                        if let Some(mfp) = desc.arguments.operators.as_ref() {
+                            stats.may_match_mfp(ResultSpec::anything(), mfp)
+                        } else {
+                            true
+                        }
+                    }
+                };
+
+                if selected {
+                    selected_bytes += bytes;
+                    selected_parts += 1u64;
+                }
+            }
+            results.push(Row::pack_slice(&[
+                name.as_str().into(),
+                total_bytes.into(),
+                selected_bytes.into(),
+                total_parts.into(),
+                selected_parts.into(),
+            ]));
+        }
+
+        Ok(results)
+    }
+
+    pub(super) async fn sequence_explain_pushdown(
+        &mut self,
+        mut ctx: ExecuteContext,
+        plan: plan::ExplainPushdownPlan,
+        target_cluster: TargetCluster,
+    ) {
+        let result = match plan.explainee {
+            Explainee::Statement(ExplaineeStatement::Select {
+                broken: false,
+                plan,
+                desc: _,
+            }) => {
+                self.explain_pushdown_select(&mut ctx, plan, target_cluster)
+                    .await
```

Comment:
Right now this will block the main loop, but I think we're moving to never running optimization on that path. Should explore other patterns here!